quinn-h3/src/connection.rs

```diff
@@ -1,51 +1,63 @@
 use std::{
     collections::{BTreeMap, HashMap, VecDeque},
-    io::Cursor,
+    io::{self, Cursor},
     mem,
     pin::Pin,
     sync::{Arc, Mutex},
     task::{Context, Waker},
 };
 
-use bytes::{Bytes, BytesMut};
+use bytes::BytesMut;
 use futures::{AsyncRead, Future, Poll, Stream};
 use quinn::{IncomingBiStreams, IncomingUniStreams, RecvStream, SendStream};
 use quinn_proto::{Side, StreamId};
 
 use crate::{
-    frame::FrameStream,
+    frame::{self, FrameStream},
     proto::{
-        connection::{Connection, DecodeResult, Error as ProtoError, PendingStreamType},
+        self,
+        connection::{
+            Connection, DecodeResult, Error as ProtoError, Error as ConnectionError,
+            PendingStreamType,
+        },
         frame::{HeadersFrame, HttpFrame},
         ErrorCode, StreamType,
     },
     streams::{NewUni, RecvUni, SendUni},
     Error, Settings,
 };
 
+struct DriverError(Error, ErrorCode, String);
+
+macro_rules! driver_err {
+    ($code:expr, $msg:expr) => {{
+        let msg = if $code == ErrorCode::INTERNAL_ERROR {
+            "".to_string()
+        } else {
+            $msg.into()
+        };
+        Err(DriverError(Error::Peer(msg.clone()), $code, msg))
+    }};
+    ($err:expr, $code:expr, $msg:expr) => {
+        Err(DriverError($err, $code, $msg.into()))
+    };
+}
 
 pub struct ConnectionDriver(pub(crate) ConnectionRef);
 
 impl Future for ConnectionDriver {
     type Output = Result<(), Error>;
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-        let mut conn = self.0.h3.lock().unwrap();
```

Comment:
Having this change in this commit seems surprising...

Reply:
I introduced a type `DriverError` to carry an `ErrorCode` along with an `Error`, the former will be sent in the quic connection shutdown, the latter will be returned to the user by the H3 driver.

This is why I introduced an extra function composition.

I could have separated the introduction of this function in a previous commit, along with `ConnectionInner::quic` removal. But it would have added work to struggle with return types in each function, twice.

Reply:
Ho, I didn't see your last comment about this. I'll get that sorted out during lunch break.